blocksuite/affine/gfx/turbo-renderer/src/turbo-renderer.ts

```diff
@@ -37,6 +43,7 @@ export class ViewportTurboRendererExtension extends LifeCycleWatcher {
 
   static override setup(di: Container) {
     di.addImpl(ViewportTurboRendererIdentifier, this, [StdIdentifier]);
+    layoutProviders.register(new ParagraphLayoutProvider());
```

Comment:
The register behavior should not be called manually. I think paragraph layout should be an extension and register itself

Reply:
Thanks for the advice, that's exactly what I need.

Just tried a few options, seems the API design with least boilerplate is to pass entities in `ConfigExtensionFactory`:

```ts
// ts
const extensions = [
  TurboRendererConfigFactory({
    // already works
    options: {
      zoomThreshold: 1,
      debounceTime: 1000,
    },
    // TODO
    blockLayoutProviders: [...],
    blockPainters: [...],
  }),
  ViewportTurboRendererExtension
]
```

In this way, each `BlockLayoutProvider` and `BlockPainter` would not be extensions (just plain class), but still can be configured and registered in downstream, with no *paragraph*-related concepts hard-coded in upstream anymore. Do you think this approach could fit the best practice?

Reply:
There are no paragraph related code in the `turbo-renderer` upstream now.

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/lEGcysB4lFTEbCwZ8jMv/3812331e-4e95-47b8-b4ea-c39a375775fe.png)

Reply:
I think the `blockLayoutProviders` and `blockLayoutPainters` can also be extension.

Reply:
Migrated, but the adoption to extension and new package structure has cost too much time this week. I thought there are still rooms to improve in terms of extension API in turbo renderer, which is better to be worked out in housekeeping stage if possible.